packages/gatsby/src/query/queue.js

```diff
@@ -14,9 +15,18 @@ const createBaseOptions = () => {
 
 const createBuildQueue = () => {
   const graphqlRunner = new GraphQLRunner(store)
+
   const handler = (queryJob, callback) =>
     queryRunner(graphqlRunner, queryJob)
-      .then(result => callback(null, result))
+      .then(result => {
+        if (process.env.GATSBY_PAGE_BUILD_ON_DATA_CHANGES && queryJob.isPage) {
+          boundActionCreators.setPageData({
+            id: queryJob.id,
+            result,
+          })
```

Comment:
This will be very expensive to keep every query result in memory. We don't do it right now. We only keep `pageContext` in memory (as part of `pages` substate)

Reply:
Does it have to be a full page data? Maybe hash will be enough?

Reply:
Yeah - and we already generate hash in https://github.com/gatsbyjs/gatsby/blob/323ad7970a88b1e067b90c60bdb83807cf66dd71/packages/gatsby/src/query/query-runner.js#L97-L100 so we could reuse that

Reply:
actually maybe change check in https://github.com/gatsbyjs/gatsby/blob/323ad7970a88b1e067b90c60bdb83807cf66dd71/packages/gatsby/src/query/query-runner.js#L102 to make use of redux state and not ad-hoc in-memory map (as it will be blown away in next session)

Reply:
I agree it's not ideal, I thought keeping it in the state over reading from each `page-data.json` directly for the data comparison would be a lesser expensive option.

Reply:
There are tradeoffs - for cases when we won't hit memory limits / keeping memory usage high won't cause excessive garbage collection keeping this in memory will be more performant than I/O reads - but with @vladar suggestion about using hash for comparisons instead of full objects - this will be handled

Reply:
Great idea about using hash.

Reply:
Nice started using the hash and seems to be working well, i'll do some more testing but looking good 🙇